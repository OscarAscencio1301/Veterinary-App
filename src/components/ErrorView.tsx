import { ReactElement } from "react"

export const ErrorView = ({ children }: { children: ReactElement }) => {
    return (
        <div aria-label="error" className="bg-red-800 text-white rounded py-2"><p>{children}</p></div>
    )
}
