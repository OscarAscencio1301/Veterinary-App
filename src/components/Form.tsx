

export const Form = () => {
  return (
    <div className="text-center">
      <h2 className="font-black text-3xl">Seguimiento</h2>
      <p className="text-lg my-5">Añade Pacientes y <span className="text-indigo-600 font-bold">Administralos</span></p>
      <form className="bg-white shadow-md rounded-md py-10 px-5">
        <div className="flex flex-col text-left gap-2 py-2">
          <label className="text-gray-700 font-bold"
            htmlFor="mascota">Nombre Mascota</label>
          <input type="text" placeholder="Nombre Mascota" id="mascota" className="p-2 border placeholder-gray-400 rounded-md" />
        </div>
        <div className="flex flex-col text-left gap-2 py-2">
          <label className="text-gray-700 font-bold"
            htmlFor="propietario">Nombre Propietario</label>
          <input type="text" placeholder="Nombre Propietario" id="propietario" className="p-2 border placeholder-gray-400 rounded-md" />
        </div>
        <div className="flex flex-col text-left gap-2 py-2">
          <label className="text-gray-700 font-bold"
            htmlFor="email">Correo Electrónico</label>
          <input type="email" placeholder="Email" id="email" className="p-2 border placeholder-gray-400 rounded-md" />
        </div>
        <div className="flex flex-col text-left gap-2 py-2">
          <label className="text-gray-700 font-bold"
            htmlFor="alta">Alta</label>
          <input type="date" id="alta" className="p-2 border placeholder-gray-400 rounded-md" />
        </div>
        <div className="flex flex-col text-left gap-2 py-2">
          <label className="text-gray-700 font-bold"
            htmlFor="sintomas">Sintomas</label>
          <textarea id="sintomas" className="p-2 border placeholder-gray-400 rounded-md" />
        </div>
        <button type="submit" className="bg-indigo-600 block w-full p-3 rounded-md text-white mt-3 hover:bg-indigo-800 cursor-pointer transition-colors duration-200">Agregar</button>
      </form>
    </div>
  )
}
