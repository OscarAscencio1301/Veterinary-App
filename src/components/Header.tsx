

export const Header = () => {
  return (
    <>
    <h1 className="font-black text-5xl text-center md:w-3/4 mx-auto">Seguimiento Paciente <span className="text-indigo-600">Veterinaria Github</span></h1>
    </>
  )
}
