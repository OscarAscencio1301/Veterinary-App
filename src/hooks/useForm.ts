import { ChangeEvent, useState } from "react"

export const useForm = <T>(initialState: T) => {

  const [form, setform] = useState(initialState)

  const changeValues = ({ target: { name, value } }: ChangeEvent<HTMLInputElement | HTMLTextAreaElement>) => {
    setform({
      ...form,
      [name]: value
    })
  }
  const reset = (newValues = initialState) => {
    setform(newValues)
  }

  return {
    ...form,
    form,
    setform,
    changeValues,
    reset
  }
}
