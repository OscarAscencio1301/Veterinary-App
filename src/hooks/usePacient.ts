import { useDispatch, useSelector } from "react-redux"
import { Pacient, Pacients, Slice } from "../interfaces/pacients"
import { active, add, clean, deleteP, update, view } from "../store/pacientSlice"



export const usePacient = () => {
  const { pacients, pacientActive } = useSelector<Slice, Pacients>(state => state.pacients)
  const dispatch = useDispatch()
  const addPacient = (pacient: Pacient) => {
    dispatch(add(pacient))
  }
  const activePacient = (pacient: Pacient) => {
    dispatch(active(pacient))
  }
  const updatePacient = (pacient: Pacient) => {
    dispatch(update(pacient))
  }
  const deletePacient = (id: string) => {
    dispatch(deleteP(id))
  }
  const cleanPacient = () => {
    dispatch(clean())
  }
  const viewPacients = (pacients:Pacient[]) => {
    dispatch(view(pacients))
  }


  return {
    pacients,
    pacientActive,
    addPacient,
    activePacient,
    cleanPacient,
    deletePacient,
    updatePacient,
    viewPacients
  }
}
