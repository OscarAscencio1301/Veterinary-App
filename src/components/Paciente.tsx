import { usePacient } from "../hooks/usePacient"
import { Pacient } from "../interfaces/pacients"

export const Paciente = ({date, email, name, pet, sintoms, id}:Pacient) => {
    
    const {deletePacient, activePacient} = usePacient()

    const deleteBTN = () => {
        deletePacient(id || '')
    }

    const updateBTN = () => {
        activePacient({date, email, name, pet, sintoms, id})
    }

    return (
        <div className="my-3 p-10 bg-white py-10 shadow-md rounded-md" aria-label="pacient">
            <p className="font-bold py-3 text-gray-700 uppercase">Mascota: <span className="font-normal normal-case">{pet}</span></p>
            <p className="font-bold py-3 text-gray-700 uppercase">Propietario: <span className="font-normal normal-case">{name}</span></p>
            <p className="font-bold py-3 text-gray-700 uppercase">Correo Electr√≥nico: <span className="font-normal normal-case">{email}</span></p>
            <p className="font-bold py-3 text-gray-700 uppercase">Alta: <span className="font-normal normal-case">{date}</span></p>
            <p className="font-bold py-3 text-gray-700 uppercase">Sintomas: <span className="font-normal normal-case">{sintoms}</span></p>
            <div className="flex justify-between">
            <button type="button" className="py-2 px-10 bg-indigo-600 hover:bg-indigo-900 rounded-md uppercase text-white font-bold" onClick={updateBTN}>Editar</button>
                <button type="button" className="py-2 px-10 bg-red-600 hover:bg-red-900 rounded-md uppercase text-white font-bold" onClick={deleteBTN}>Eliminar</button>
            </div>
        </div>
    )
}
