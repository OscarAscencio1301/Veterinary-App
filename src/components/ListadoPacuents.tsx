import { Paciente } from "./Paciente"

export const ListadoPacientes = () => {
  return (
    <div className="text-center col-span-2 h-screen overflow-y-auto">
      <h2 className="font-black text-3xl">Listado de Pacientes</h2>
      <p className="text-lg my-5">Administra tus <span className="text-indigo-600 font-bold">Pacientes y Citas.</span></p>
      <Paciente />
      <Paciente />
      <Paciente />
    </div>
  )
}
