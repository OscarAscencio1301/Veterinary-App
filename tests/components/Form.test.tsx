import { fireEvent, render, screen } from "@testing-library/react";
import { Form } from "../../src/components/Form";
import { useForm } from "../../src/hooks/useForm";
import { usePacient } from "../../src/hooks/usePacient";

jest.mock("../../src/hooks/useForm");
jest.mock("../../src/hooks/usePacient");

describe('Testing Component Form', () => {
    const addPacientMock = jest.fn()
    const resetMock = jest.fn()
    beforeEach(() => jest.clearAllMocks())


    test('Snapshot ', () => {
        (useForm as jest.Mock).mockReturnValue({
            changeValues: jest.fn(),
            reset: resetMock,
            setform: jest.fn()
        });

        (usePacient as jest.Mock).mockReturnValue({
            addPacient: jest.fn(),
            cleanPacient: jest.fn(),
            updatePacient: jest.fn(),
            pacientActive: null
        });

        const { container } = render(<Form />)
        expect(container).toMatchSnapshot()
    });
    test('Initial Values ', () => {
        (useForm as jest.Mock).mockReturnValue({
            changeValues: jest.fn(),
            reset: resetMock,
            setform: jest.fn()
        });

        (usePacient as jest.Mock).mockReturnValue({
            addPacient: jest.fn(),
            cleanPacient: jest.fn(),
            updatePacient: jest.fn(),
            pacientActive: null
        });

        render(<Form />)
        expect(screen.getAllByRole('textbox').length).toBe(4)
    });
    test('Change Values in Inputs', () => {
        
        (useForm as jest.Mock).mockReturnValue({
            changeValues: jest.fn(),
            setform: jest.fn()
        });

        (usePacient as jest.Mock).mockReturnValue({
            addPacient: jest.fn(),
            cleanPacient: jest.fn(),
            updatePacient: jest.fn(),
            pacientActive: null
        });

        render(<Form />)
        const inputs = screen.getAllByRole<HTMLInputElement>('textbox')
        const date = screen.getByLabelText<HTMLInputElement>('date')
        const textarea = screen.getByLabelText<HTMLTextAreaElement>('textarea')

        fireEvent.input(inputs[0], { target: { value: 'Perrinho' } })
        fireEvent.input(inputs[1], { target: { value: 'Oscar Test' } })
        fireEvent.input(inputs[2], { target: { value: 'oscar@oscar.com' } })
        fireEvent.input(date, { target: { value: '2023-07-13' } })
        fireEvent.input(textarea, { target: { value: 'Le duele la pata derecha' } })


        expect(inputs[0].value).toBe('Perrinho')
        expect(inputs[1].value).toBe('Oscar Test')
        expect(inputs[2].value).toBe('oscar@oscar.com')
        expect(date.value).toBe('2023-07-13')
        expect(textarea.value).toBe('Le duele la pata derecha')
    });
    test('Submit', () => {

        (useForm as jest.Mock).mockReturnValue({
            changeValues: jest.fn(),
            reset: resetMock,
            setform: jest.fn()
        });

        (usePacient as jest.Mock).mockReturnValue({
            addPacient: addPacientMock,
            cleanPacient: jest.fn(),
            updatePacient: jest.fn(),
            pacientActive: null
        });

        render(<Form />)
        const inputs = screen.getAllByRole<HTMLInputElement>('textbox')
        const date = screen.getByLabelText<HTMLInputElement>('date')
        const textarea = screen.getByLabelText<HTMLTextAreaElement>('textarea')
        const form = screen.getByLabelText('form')

        fireEvent.input(inputs[0], { target: { value: 'Perrinho' } })
        fireEvent.input(inputs[1], { target: { value: 'Oscar Test' } })
        fireEvent.input(inputs[2], { target: { value: 'oscar@oscar.com' } })
        fireEvent.input(date, { target: { value: '2023-07-13' } })
        fireEvent.input(textarea, { target: { value: 'Le duele la pata derecha' } })
        fireEvent.submit(form)

        expect(addPacientMock).toBeCalled()
        expect(resetMock).toBeCalled()
    });
});