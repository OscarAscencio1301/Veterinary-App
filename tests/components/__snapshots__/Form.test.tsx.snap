// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Component Form Snapshot  1`] = `
<div>
  <div
    class="text-center"
  >
    <h2
      class="font-black text-3xl"
    >
      Seguimiento
    </h2>
    <p
      class="text-lg my-5"
    >
      Añade Pacientes y 
      <span
        class="text-indigo-600 font-bold"
      >
        Administralos
      </span>
    </p>
    <form
      aria-label="form"
      class="bg-white shadow-md rounded-md py-10 px-5"
    >
      <div
        class="flex flex-col text-left gap-2 py-2"
      >
        <label
          class="text-gray-700 font-bold"
          for="mascota"
        >
          Nombre Mascota
        </label>
        <input
          class="p-2 border placeholder-gray-400 rounded-md"
          id="mascota"
          name="pet"
          placeholder="Nombre Mascota"
          type="text"
          value=""
        />
      </div>
      <div
        class="flex flex-col text-left gap-2 py-2"
      >
        <label
          class="text-gray-700 font-bold"
          for="propietario"
        >
          Nombre Propietario
        </label>
        <input
          class="p-2 border placeholder-gray-400 rounded-md"
          id="propietario"
          name="name"
          placeholder="Nombre Propietario"
          type="text"
          value=""
        />
      </div>
      <div
        class="flex flex-col text-left gap-2 py-2"
      >
        <label
          class="text-gray-700 font-bold"
          for="email"
        >
          Correo Electrónico
        </label>
        <input
          class="p-2 border placeholder-gray-400 rounded-md"
          id="email"
          name="email"
          placeholder="Email"
          type="email"
          value=""
        />
      </div>
      <div
        class="flex flex-col text-left gap-2 py-2"
      >
        <label
          class="text-gray-700 font-bold"
          for="alta"
        >
          Alta
        </label>
        <input
          aria-label="date"
          class="p-2 border placeholder-gray-400 rounded-md"
          id="alta"
          name="date"
          type="date"
          value=""
        />
      </div>
      <div
        class="flex flex-col text-left gap-2 py-2"
      >
        <label
          class="text-gray-700 font-bold"
          for="sintomas"
        >
          Sintomas
        </label>
        <textarea
          aria-label="textarea"
          class="p-2 border placeholder-gray-400 rounded-md"
          id="sintomas"
          name="sintoms"
        />
      </div>
      <button
        class="bg-indigo-600 block w-full p-3 rounded-md text-white mt-3 hover:bg-indigo-800 cursor-pointer transition-colors duration-200"
        type="submit"
      >
        Agregar Paciente
      </button>
    </form>
  </div>
</div>
`;
